Quarkus langchain4j blog post (#1833)
quarkusio/quarkusio.github.io: _posts/2023-11-16-quarkus-meets-langchain4j.adoc

---
layout: post
title: 'When Quarkus meets LangChain4j'
date: 2023-11-15
tags: AI langchain
synopsis: 'Learn about the new quarkus-langchain4j extension to integrate LLMs in Quarkus applications.'
author: cescoffier
---
:imagesdir: /assets/images/posts/llm

Large language models (LLMs) are reshaping the world of software, altering the way we interact with users and develop business logic.

Populatized by https://openai.com/[OpenAI]'s https://chat.openai.com/[ChatGPT], LLMs are now available in many flavors and sizes. The https://huggingface.co/models[Hugging-Face] platform references hundreds of them, and major tech companies like Facebook, Google, Microsoft, and Amazon are also providing their own models.

> The Quarkus team, in collaboration with the langchain4j team, has been working on an extension to integrate LLMs in Quarkus applications. This extension is based on the https://github.com/langchain4j[langchain4j library], which provides a common API to interact with LLMs. The langchain4j project is a Java re-implementation of the famous https://www.langchain.com/[langchain] library.

Reviewer: ```suggestion
The Quarkus team, in collaboration with the Langchain4j team, has been working on an extension to integrate LLMs in Quarkus applications. This extension is based on the https://github.com/langchain4j[langchain4j library], which provides a common API to interact with LLMs. The langchain4j project is a Java re-implementation of the famous https://www.langchain.com/[langchain] library.
```

Strictly relying on the LLM's knowledge might not be enough. Thus, the Quarkus Langchain4j extension provides two mechanisms to extend AI knowledge:

- 1) Tools - a tool lets the LLM execute actions in your application. For instance, you can use a tool to send an email, call a REST endpoint, or execute a database query. The LLM decides when to use the tool, the method parameters, and what to do with the result.
- 2) Document stores - LLMs are not good at remembering things. In addition, their context has a size limit. Thus, the extension provides a way to store and retrieve information from document stores. Before calling the LLM, the extension can ask for relevant documents in a document store and attach them to the context. The LLM can then use this data to make a decision. For instance, you can load spreadsheet data, reports, or data from a database.

The following diagram illustrates the interactions between the LLM, the tools, and the document stores:

// TODO: Add diagram here

== Show me some code!

> Alright, enough "bla bla," let's see some code. We are going to use Open AI gpt-3.5 (be careful that it's not the state-of-the-art model, but it's good enough for this demo), give it some product reviews, and ask the LLM to classify them between positive and negative reviews. The full code is available in the [quarkus-langchain4j repository](https://github.com/quarkiverse/quarkus-langchain4j/tree/main/samples/review-triage).

Reviewer: ```suggestion
Alright, enough "bla bla," let's see some code! We are going to use Open AI gpt-3.5 (be careful that it's not the state-of-the-art model, but it's good enough for this demo), give it some product reviews, and ask the LLM to classify them between positive and negative reviews. The full code is available in the [quarkus-langchain4j repository](https://github.com/quarkiverse/quarkus-langchain4j/tree/main/samples/review-triage).
```

[source, java]
----
@RegisterAiService
public interface TriageService {
    // methods.
}
----

The rest of the application would be able to use the LLM by injecting the `TriageService` interface and calling the methods.

Speaking about methods, that's where the magic happens. You will describe what you want the LLM to do using natural language. First, you start with `@SystemMessage` to define the role and scope. Then, you can use `@UserMessage` to describe the task.

[source, java]
----
@SystemMessage("""
    You are working for a bank, processing reviews about financial products. Triage reviews into positive and negative ones, responding with a JSON document.
    """
)
@UserMessage("""
    Your task is to process the review delimited by ---.
    Apply sentiment analysis to the review to determine if it is positive or negative, considering various languages.

    For example:
    - "I love your bank, you are the best!" is a 'POSITIVE' review
    - "J'adore votre banque" is a 'POSITIVE' review

>     - "I hate your bank, you are the worst!" is a 'NEGATIVE' review

Reviewer: Why not use Java's multiline syntax?

== Tools and Document loader

The previous example is a bit simplistic. In the real world, you will need to extend the LLM knowledge with tools and document stores. The `@RegisterAiService` annotation lets you define the tools and document stores to use.

=== Tools

Tools are methods that the LLM can invoke. Note that not all LLMs support tools.

To declare a tool, just use the `@Tool` annotation in a bean method:

[source, java]
----
@ApplicationScoped
public class CustomerRepository implements PanacheRepository<Customer> {

    @Tool("get the customer name for the given customerId")
    public String getCustomerName(long id) {
        return find("id", id).firstResult().name;
    }

}
----

In this example, we are using the Panache repository pattern to access the database. We have a specific method annotated with `@Tool` to retrieve the customer name. When the LLM needs to get the customer name, it instructs Quarkus to call this method and receives the result.

Obviously, it's not a good idea to expose every operation to the LLM. So, in addition to `@Tool`, you need to list the set of tools you allow the LLM to invoke in the `@RegisterAiService` annotation:

[source, java]
----

> @RegisterAiService(tools = { TransactionRepository.class, CustomerRepository.class })

Reviewer: Don't we also need to configure the memory?

Author: you are right! Forgot about that.

public interface FraudDetectionAi {
   // ...
}
----

=== Document stores

> Document stores are a way to extend the LLM knowledge with your own data. This approach named _RAG_ requires two processes:

Reviewer: ```suggestion
Document stores are a way to extend the LLM knowledge with your own data. This approach - called  _RAG_ - requires two processes:
```

First, you need to ingest documents into a document store. The documents are not stored as-is, but an embedding is computed. This embedding is a vector representation of the document.

> Then, in the Quarkus application, you need to declare the document store and the embedding to use. The extension will load the document store and the embedding and will use them to extend the LLM knowledge. Thus, before calling the LLM, it retrieves the relevant documents from the store (that's where the vector representation is useful) and attaches them to the LLM context.

Reviewer: ```suggestion
Then, in the Quarkus application, you need to declare the document store and the embedding to use. The extension will load the document store and the embedding and will use them to extend the LLM knowledge. Thus, before calling the LLM, it retrieves the relevant documents from the store (that's where the vector representation is useful) and attaches them to the LLM context (which essentially means adding the retrieved information from the document to the user message).
```

---
layout: post
title: 'When Quarkus meets LangChain4j'
date: 2023-11-15
tags: AI langchain
synopsis: 'Learn about the new quarkus-langchain4j extension to integrate LLMs in Quarkus applications.'
author: cescoffier
---
:imagesdir: /assets/images/posts/llm

Large language models (LLMs) are reshaping the world of software, altering the way we interact with users and develop business logic.

Populatized by https://openai.com/[OpenAI]'s https://chat.openai.com/[ChatGPT], LLMs are now available in many flavors and sizes. The https://huggingface.co/models[Hugging-Face] platform references hundreds of them, and major tech companies like Facebook, Google, Microsoft, and Amazon are also providing their own models.

> LLMs are not a new concept. They have been around for a while, but they were not as powerful as they are today. The Quarkus team have been thinking about what it would mean to integrate LLMs in Quarkus applications. The talk https://www.youtube.com/watch?app=desktop&v=BD1MSLbs9KE[Java Meets AI] from Lize Raes at Devoxx 2023 has been a great source of inspiration.

Reviewer 2: ```suggestion
LLMs are not a new concept. They have been around for a while, but they were not as powerful or as accessible the became when OpenAI made ChatGPT API's publically available. Since then the Quarkus team have been thinking about what it would mean to integrate LLMs in the Quarkus ecosystem. The talk https://www.youtube.com/watch?app=desktop&v=BD1MSLbs9KE[Java Meets AI] from Lize Raes at Devoxx 2023 has been a great source of inspiration.
```

> Since, the Quarkus team, in collaboration with the Langchain4j team, has been working on an extension to integrate LLMs in Quarkus applications. This extension is based on the https://github.com/langchain4j[Langchain4j library], which provides a common API to interact with LLMs. The Langchain4j project is a Java re-implementation of the famous https://www.langchain.com/[langchain] library.

Reviewer 2: ```suggestion
Since, the Quarkus team, in collaboration with the LangChain4j team, has been working on an extension to integrate LLMs in Quarkus applications. This extension is based on the https://github.com/langchain4j[Langchain4j library], which provides a common API to interact with LLMs. The Langchain4j project is a Java re-implementation of the famous https://www.langchain.com/[langchain] library.
```

Reviewer 2: I'm pondering how we can phrase this to highlight that this is *one* way of integrating with LLM's and that we know the AI space are moving fast so not clear yet what wins - thus are inviting others to contribute to this extension or build others.

Author: I tried to say this in the conclusion.

Reviewer 2: yeah - trying to have it hinted up front early as some might not ever read further :)

First, let's have a look at the big picture. When integrating an LLM into a Quarkus application, you need to describe what you want the AI to do. Unlike traditional code, you are going to explain the behavior of the AI using natural language. Of course, there are a few techniques to tame the AI, but we will explore that later.

Strictly relying on the LLM's knowledge might not be enough. Thus, the Quarkus Langchain4j extension provides two mechanisms to extend AI knowledge:

- 1) Tools - a tool lets the LLM execute actions in your application. For instance, you can use a tool to send an email, call a REST endpoint, or execute a database query. The LLM decides when to use the tool, the method parameters, and what to do with the result.
- 2) Document stores - LLMs are not good at remembering things. In addition, their context has a size limit. Thus, the extension provides a way to store and retrieve information from document stores. Before calling the LLM, the extension can ask for relevant documents in a document store and attach them to the context. The LLM can then use this data to make a decision. For instance, you can load spreadsheet data, reports, or data from a database.

> 

Reviewer 2: this is all true but I feel like we are repeating the same approach I feel others communicating about AI - introducing it using new terms and approaches when 90% of this is truly just simple stuff and that lots of existing java and quarkus functionality fits well into it.

Reviewer 3: Not sure about this. I read the doc the other day (which presents the same things) and I found it extremely informative and useful to put names on concepts.

Reviewer 2: note: I wasn't referring to Document stores specifically - but its not a big deal after rereading it - it is using more "normal" words than others i've seen so +1.

=== Tools

Tools are methods that the LLM can invoke.

To declare a tool, just use the `@Tool` annotation in a bean method:

[source, java]
----
@ApplicationScoped
public class CustomerRepository implements PanacheRepository<Customer> {

    @Tool("get the customer name for the given customerId")
    public String getCustomerName(long id) {
        return find("id", id).firstResult().name;
    }

}
----

In this example, we are using the Panache repository pattern to access the database. We have a specific method annotated with `@Tool` to retrieve the customer name. When the LLM needs to get the customer name, it instructs Quarkus to call this method and receives the result.

Obviously, it's not a good idea to expose every operation to the LLM. So, in addition to `@Tool`, you need to list the set of tools you allow the LLM to invoke in the `@RegisterAiService` annotation:

[source, java]
----
@RegisterAiService(tools = { TransactionRepository.class, CustomerRepository.class })
public interface FraudDetectionAi {
   // ...
}
----

>  Note that not all LLMs support tools.

Reviewer 2: should we mention somewhere which LLM's that are currently supported/working with quarkus-langchain4j?

Author: only openai unfortunately. I didn't find a HF model supporting tools yet (but only tried 5 of them)

Author: I don't know about ollama (@geoand do you know)?

Reviewer: My guess is that it does not as the REST request / response does not contain anything indicating such a possibility

=== Document stores

> Document stores are a way to extend the LLM knowledge with your own data. This approach - called  _RAG_ - requires two processes:

Reviewer 2: ```suggestion
Document stores are a way to extend the LLM knowledge with your own data. This approach - called  Retrieval Augmented Generation (_RAG_) - requires two processes:
```

Technically it is not RAG to extend with own data - the RAG part is selectively extract a sub set of your data to send to LLM to reduce noise and overhead.

Author: it's only the second step. The first step is the ingestion.

                .documentSplitter(recursive(500, 0))
                .build();  
        ingestor.ingest(documents);
    }
}
----

Then, generally, in another application, you can use the populated document store to extend the LLM knowledge. First, create a bean implementing the `Retriever<TextSegment>` interface:

[source, java]
----
@ApplicationScoped
public class DocumentRetrieverExample implements Retriever<TextSegment> {

    private final EmbeddingStoreRetriever retriever;

    DocumentRetrieverExample(RedisEmbeddingStore store, EmbeddingModel model) {
        retriever = EmbeddingStoreRetriever.from(store, model, 20);
    }

    @Override
    public List<TextSegment> findRelevant(String s) {
        return retriever.findRelevant(s);
    }
}
----

Then, add the document store and the retriever to the `@RegisterAiService` annotation:

[source, java]
----
@RegisterAiService(retrieverSupplier = RegisterAiService.BeanRetrieverSupplier.class)
public interface MyAiService {
// ...
}
----

TIP: `RegisterAiService.BeanRetrieverSupplier.class` is a special value looking for the `Retriever` bean in the Quarkus application.

== Final notes

> This post presented the Quarkus Langchain4j extension. This is the first version of the extension, and we continue exploring what integrating LLMs into Quarkus applications means. We are looking for feedback and ideas to improve the extension. We are working removing some rough angles. 

Reviewer 2: I would add some obvious but maybe not so obvious advantages this level of integration into quarkus has and what it enables:

This all fits nicely with CDI - you can basically hook up any of your CDI beans as tools or as feed into data. Camel routes can be beans, camel can use a AIService in its routes, etc.

> This post presented the Quarkus Langchain4j extension. This is the first version of the extension, and we continue exploring what integrating LLMs into Quarkus applications means. We are looking for feedback and ideas to improve the extension. We are working removing some rough angles. 

Reviewer 2: Maybe highlight that this extension is an experiment to see what is possible. The approach will evolve and inviting other LLM libraries to contribute to other extensions and explore ways to bring developer joy to LLM apps.

Since, the Quarkus team, in collaboration with the Langchain4j team, has been working on an extension to integrate LLMs in Quarkus applications. This extension is based on the https://github.com/langchain4j[Langchain4j library], which provides a common API to interact with LLMs. The Langchain4j project is a Java re-implementation of the famous https://www.langchain.com/[langchain] library.

In this blog post, we will see how to use the https://docs.quarkiverse.io/quarkus-langchain4j/dev/index.html[quarkus-langchain4j] extension to integrate LLMs in Quarkus applications. This extension is an exploration to understand how LLMs can be used in Quarkus applications.

== Overview

First, let's have a look at the big picture. When integrating an LLM into a Quarkus application, you need to describe what you want the AI to do. Unlike traditional code, you are going to explain the behavior of the AI using natural language. Of course, there are a few techniques to tame the AI, but we will explore that later.

> Strictly relying on the LLM's knowledge might not be enough. Thus, the Quarkus Langchain4j extension provides two mechanisms to extend AI knowledge:

Reviewer 3: ```suggestion
Strictly relying on the LLM's knowledge might not be enough. Thus, the Quarkus LangChain4j extension provides two mechanisms to extend AI knowledge:
```

- 1) Tools - a tool lets the LLM execute actions in your application. For instance, you can use a tool to send an email, call a REST endpoint, or execute a database query. The LLM decides when to use the tool, the method parameters, and what to do with the result.
- 2) Document stores - LLMs are not good at remembering things. In addition, their context has a size limit. Thus, the extension provides a way to store and retrieve information from document stores. Before calling the LLM, the extension can ask for relevant documents in a document store and attach them to the context. The LLM can then use this data to make a decision. For instance, you can load spreadsheet data, reports, or data from a database.

The following diagram illustrates the interactions between the LLM, the tools, and the document stores:

// TODO: Add diagram here

== Show me some code!

> Alright, enough "bla bla," let's see some code! We are going to use Open AI GPT-3.5 (be careful that it's not the state-of-the-art model, but it's good enough for this demo), give it some product reviews, and ask the LLM to classify them between positive and negative reviews. The full code is available in the [quarkus-langchain4j repository](https://github.com/quarkiverse/quarkus-langchain4j/tree/main/samples/review-triage).

Reviewer 3: ```suggestion
Alright, enough "bla bla", let's see some code! We are going to use Open AI GPT-3.5 (be careful that it's not the state-of-the-art model, but it's good enough for this demo), give it some product reviews, and ask the LLM to classify them between positive and negative reviews. The full code is available in the [quarkus-langchain4j repository](https://github.com/quarkiverse/quarkus-langchain4j/tree/main/samples/review-triage).
```

First, we need the `quarkus-langchain4j-openai` extension:

[source, xml]
----
<dependency>
    <groupId>io.quarkiverse.langchain4j</groupId>
    <artifactId>quarkus-langchain4j-openai</artifactId>
    <version>0.1.0</version> <!-- Update to use the latest version -->
</dependency>
----

Once we have the extension, it's time to tell the LLM what we want to do. The Quarkus LangChain4J extension provides a declarative way to describe LLM interactions. The idea is the same as the Quarkus REST client. We model the interaction using an interface annotated with `@RegisterAiService`:

[source, java]
----
@RegisterAiService
public interface TriageService {
    // methods.
}
----

The rest of the application would be able to use the LLM by injecting the `TriageService` interface and calling the methods.

Speaking about methods, that's where the magic happens. You will describe what you want the LLM to do using natural language. First, you start with `@SystemMessage` to define the role and scope. Then, you can use `@UserMessage` to describe the task.

[source, java]
----

> @SystemMessage("""

Reviewer 3: I think I would keep the `class` part. It doesn't take much space and makes things a bit clearer. But YMMV.

[source, java]
----
@Path("/review")
public class ReviewResource {

    @Inject
    TriageService triage;

    record Review(String review) {
      // User text
    }

    @POST
    public TriagedReview triage(Review review) {
        return triage.triage(review.review());
    }

}
----

That's it! The LLM is now integrated into the application. The `TriageService` interface is used as an ambassador to call the LLM. This declarative approach has many advantages:

- Testability - you can easily mock the LLM by providing a fake implementation of the interface.
- Observability - you can use the Quarkus metrics annotation to monitor the LLM methods.
- Resilience - you can use the Quarkus fault-tolerance annotations to handle failures, timeouts, and other transient issues.

== Tools and Document loader

The previous example is a bit simplistic. In the real world, you will need to extend the LLM knowledge with tools and document stores. The `@RegisterAiService` annotation lets you define the tools and document stores to use.

=== Tools

Tools are methods that the LLM can invoke.

> To declare a tool, just use the `@Tool` annotation in a bean method:

Reviewer 3: ```suggestion
To declare a tool, just use the `@Tool` annotation on a bean method:
```

[source, java]
----
@ApplicationScoped
public class CustomerRepository implements PanacheRepository<Customer> {

    @Tool("get the customer name for the given customerId")
    public String getCustomerName(long id) {
        return find("id", id).firstResult().name;
    }

}
----

In this example, we are using the Panache repository pattern to access the database. We have a specific method annotated with `@Tool` to retrieve the customer name. When the LLM needs to get the customer name, it instructs Quarkus to call this method and receives the result.

Obviously, it's not a good idea to expose every operation to the LLM. So, in addition to `@Tool`, you need to list the set of tools you allow the LLM to invoke in the `@RegisterAiService` annotation:

[source, java]
----
@RegisterAiService(
    tools = { TransactionRepository.class, CustomerRepository.class },
    chatMemoryProviderSupplier = RegisterAiService.BeanChatMemoryProviderSupplier.class
)
public interface FraudDetectionAi {
   // ...
}
----

> The `chatMemoryProviderSupplier` configuration may raise questions. When using tools, a sequence of messages unfolds behind the scenes. It becomes necessary to configure the AI service's memory to adeptly track these interactions. The `chatMemoryProviderSupplier` offers a solution by enabling you to define how the memory is handled. The `chatMemoryProviderSupplier` lets you define how the memory is managed. The `BeanChatMemoryProviderSupplier` looks for a `ChatMemoryProvider` bean, like the following:

Reviewer 3: > The `chatMemoryProviderSupplier` offers a solution by enabling you to define how the memory is handled. The `chatMemoryProviderSupplier` lets you define how the memory is managed.

These two sentences are a bit repetitive. Is it normal?

    @Override
    public ChatMemory get(Object memoryId) {
        return memories.computeIfAbsent(memoryId, id -> MessageWindowChatMemory.builder()
                .maxMessages(20)
                .id(memoryId)
                .build());
    }

    @PreDestroy
    public void close() {
        memories.clear();
    }
}
----

IMPORTANT: Note that not all LLMs support tools.

=== Document stores

Document stores are a way to extend the LLM knowledge with your own data. This approach - called  _RAG_ - requires two processes:

First, you need to ingest documents into a document store. The documents are not stored as-is, but an embedding is computed. This embedding is a vector representation of the document.

Then, in the Quarkus application, you need to declare the document store and the embedding to use. The extension will load the document store and the embedding and will use them to extend the LLM knowledge. Thus, before calling the LLM, it retrieves the relevant documents from the store (that's where the vector representation is useful) and attaches them to the LLM context (which essentially means adding the retrieved information from the document to the user message).

> The Quarkus Langchain4j extension provides facilities for both steps.

Reviewer 3: ```suggestion
The Quarkus LangChain4j extension provides facilities for both steps.
```

The following code shows how to ingest a document into a Redis document store:

[source, java]
----
@ApplicationScoped

> public class IngestorExample {

Reviewer 3: Could this  be linked to the previous example somehow?

Author: WDYM?

Reviewer 3: Well, we have a very good example above that allows to understand exactly why we have these concepts and this code here is disconnected from that so it's harder to imagine how it could be useful.

But I'm not sure if we could have a usage for this concept in the previous triaging example.

                .documentSplitter(recursive(500, 0))
                .build();  
        ingestor.ingest(documents);
    }
}
----

Then, generally, in another application, you can use the populated document store to extend the LLM knowledge. First, create a bean implementing the `Retriever<TextSegment>` interface:

[source, java]
----
@ApplicationScoped
public class DocumentRetrieverExample implements Retriever<TextSegment> {

    private final EmbeddingStoreRetriever retriever;

    DocumentRetrieverExample(RedisEmbeddingStore store, EmbeddingModel model) {
        retriever = EmbeddingStoreRetriever.from(store, model, 20);
    }

    @Override
    public List<TextSegment> findRelevant(String s) {
        return retriever.findRelevant(s);
    }
}
----

Then, add the document store and the retriever to the `@RegisterAiService` annotation:

[source, java]
----
@RegisterAiService(retrieverSupplier = RegisterAiService.BeanRetrieverSupplier.class)
public interface MyAiService {
// ...
}
----

TIP: `RegisterAiService.BeanRetrieverSupplier.class` is a special value looking for the `Retriever` bean in the Quarkus application.

== Final notes

> This post presented the Quarkus Langchain4j extension. This is the first version of the extension, and we continue exploring what integrating LLMs into Quarkus applications means. We are looking for feedback and ideas to improve the extension. We are working removing some rough angles. 

Reviewer 3: ```suggestion
This post presented the Quarkus LangChain4j extension. This is the first version of the extension, and we continue exploring what integrating LLMs into Quarkus applications means. We are looking for feedback and ideas to improve the extension. We are working removing some rough angles. 
```

> This extension would not be possible without the fantastic work from Dmytro Liubarskyi in the Langchain4j library. Our collaboration has allowed us to provide a Quarkus-friendly approach to integrate the library (including native compilation support) and shapes a new way to integrate LLMs in Quarkus applications.

Reviewer 3: ```suggestion
This extension would not be possible without the fantastic work from Dmytro Liubarskyi on the LangChain4j library. Our collaboration has allowed us to provide a Quarkus-friendly approach to integrate the library (including native compilation support) and shapes a new way to integrate LLMs in Quarkus applications.
```

Popularized by https://openai.com/[OpenAI]'s https://chat.openai.com/[ChatGPT], LLMs are now available in many flavors and sizes. The https://huggingface.co/models[Hugging-Face] platform references hundreds of them, and major tech companies like Facebook, Google, Microsoft, Amazon and IBM are also providing their own models.

LLMs are not a new concept. They have been around for a while, but they were not as powerful or as accessible the became when OpenAI made ChatGPT API's publically available. Since then the Quarkus team have been thinking about what it would mean to integrate LLMs in the Quarkus ecosystem. The talk https://www.youtube.com/watch?app=desktop&v=BD1MSLbs9KE[Java Meets AI] from Lize Raes at Devoxx 2023 has been a great source of inspiration.

Since, the Quarkus team, in collaboration with the LangChain4j team, has been working on an extension to integrate LLMs in Quarkus applications. This extension is based on the https://github.com/langchain4j[LangChain4j library], which provides a common API to interact with LLMs. The LangChain4j project is a Java re-implementation of the famous https://www.langchain.com/[langchain] library.

> In this blog post, we will see how to use the https://docs.quarkiverse.io/quarkus-langchain4j/dev/index.html[quarkus-langchain4j] extension to integrate LLMs in Quarkus applications. This extension is an exploration to understand how LLMs can be used in Quarkus applications.

Reviewer 2: ```suggestion
In this blog post, we will see how to use the just released https://docs.quarkiverse.io/quarkus-langchain4j/dev/index.html[quarkus-langchain4j] 0.1 extension to integrate LLMs in Quarkus applications. This extension is an exploration to understand how LLMs can be used in Quarkus applications.
```

Strictly relying on the LLM's knowledge might not be enough. Thus, the Quarkus LangChain4j extension provides two mechanisms to extend AI knowledge:

- 1) Tools - a tool lets the LLM execute actions in your application. For instance, you can use a tool to send an email, call a REST endpoint, or execute a database query. The LLM decides when to use the tool, the method parameters, and what to do with the result.
- 2) Document stores - LLMs are not good at remembering things. In addition, their context has a size limit. Thus, the extension provides a way to store and retrieve information from document stores. Before calling the LLM, the extension can ask for relevant documents in a document store and attach them to the context. The LLM can then use this data to make a decision. For instance, you can load spreadsheet data, reports, or data from a database.

The following diagram illustrates the interactions between the LLM, the tools, and the document stores:

// TODO: Add diagram here

== Show me some code!

> Alright, enough "bla bla", let's see some code! We are going to use Open AI GPT-3.5 (be careful that it's not the state-of-the-art model, but it's good enough for this demo), give it some product reviews, and ask the LLM to classify them between positive and negative reviews. The full code is available in the [quarkus-langchain4j repository](https://github.com/quarkiverse/quarkus-langchain4j/tree/main/samples/review-triage).

Reviewer 2: ```suggestion
Alright, enough "bla bla", let's see some code! We are going to use Open AI GPT-3.5 (be careful that it's not the state-of-the-art model, but it's good enough for this demo), give it some product reviews, and ask the LLM to classify them between positive and negative reviews. The full code is available in the https://github.com/quarkiverse/quarkus-langchain4j/tree/main/samples/review-triage[quarkus-langchain4j repository].
```

---
layout: post
title: 'When Quarkus meets LangChain4j'
date: 2023-11-15
tags: AI langchain
synopsis: 'Learn about the new quarkus-langchain4j extension to integrate LLMs in Quarkus applications.'
author: cescoffier
---
:imagesdir: /assets/images/posts/llms

Large language models (LLMs) are reshaping the world of software, altering the way we interact with users and develop business logic.

Popularized by https://openai.com/[OpenAI]'s https://chat.openai.com/[ChatGPT], LLMs are now available in many flavors and sizes. The https://huggingface.co/models[Hugging-Face] platform references hundreds of them, and major tech companies like Facebook, Google, Microsoft, Amazon and IBM are also providing their own models.

> LLMs are not a new concept. They have been around for a while, but they were not as powerful or as accessible the became when OpenAI made ChatGPT API's publically available. Since then the Quarkus team have been thinking about what it would mean to integrate LLMs in the Quarkus ecosystem. The talk https://www.youtube.com/watch?app=desktop&v=BD1MSLbs9KE[Java Meets AI] from Lize Raes at Devoxx 2023 has been a great source of inspiration.

Reviewer 4: **the** became when OpenAI made ChatGPT -> **they** became when OpenAI made ChatGPT

Reviewer: PRs are open :)

Author: Max has a PR opened, I will add to MAx's PR.

> Since, the Quarkus team, in collaboration with Dmytro Liubarskyi and the LangChain4j team, has been working on an extension to integrate LLMs in Quarkus applications. This extension is based on the https://github.com/langchain4j[LangChain4j library], which provides a common API to interact with LLMs. The LangChain4j project is a Java re-implementation of the famous https://www.langchain.com/[langchain] library.

Reviewer 4: **Since** -> **Since then** maybe? Although it was already used in the paragraph above....

In this blog post, we will see how to use the just released https://docs.quarkiverse.io/quarkus-langchain4j/dev/index.html[quarkus-langchain4j] 0.1 extension to integrate LLMs in Quarkus applications. This extension is an exploration to understand how LLMs can be used in Quarkus applications.

== Overview

First, let's have a look at the big picture. When integrating an LLM into a Quarkus application, you need to describe what you want the AI to do. Unlike traditional code, you are going to explain the behavior of the AI using natural language. Of course, there are a few techniques to tame the AI, but we will explore that later.

Strictly relying on the LLM's knowledge might not be enough. Thus, the Quarkus LangChain4j extension provides two mechanisms to extend AI knowledge:

> - 1) Tools - a tool lets the LLM execute actions in your application. For instance, you can use a tool to send an email, call a REST endpoint, or execute a database query. The LLM decides when to use the tool, the method parameters, and what to do with the result.

Reviewer 4: It should be probably an ordered list, so maybe replace `-` with `.`  and remove `1)`?
